File: ios/Sources/MoQKit/Subscribe/PlayerEvents.swift
import Foundation

// MARK: - PlayerEvent

/// Stable event names emitted by ``Player/subscribeEvents(_:)``.
///
/// Events represent transitions in the player's lifecycle. Periodic samples (bitrate,
/// latency, buffer fill, fps) are not events — subscribe to ``Player/subscribeStats(_:)``
/// for those.
public enum PlayerEventName: String, Sendable {
    case playerInit = "player.init"
    case playerDestroy = "player.destroy"
    case playbackRequest = "playback.request"
    case playbackStart = "playback.start"
    case playbackPause = "playback.pause"
    case playbackResume = "playback.resume"
    case playbackEnd = "playback.end"
    case trackSubscribeStart = "track.subscribe.start"
    case trackReady = "track.ready"
    case trackPlaying = "track.playing"
    case trackSubscribeError = "track.subscribe.error"
    case trackSubscribeEnd = "track.subscribe.end"
    case trackSelect = "track.select"
    case trackSwitch = "track.switch"
    case trackStallStart = "track.stall.start"
    case trackStallEnd = "track.stall.end"
    case rebufferStart = "rebuffer.start"
    case rebufferEnd = "rebuffer.end"
    case decodeError = "decode.error"
}

/// Media kind carried by player events.
public enum PlayerTrackKind: String, Sendable {
    case audio
    case video
}

/// Session-level context shared by player lifecycle events.
public struct PlayerSessionEvent: Sendable {
    public let catalogPath: String
    public let targetBuffering: Duration
    public let videoTrackName: String?
    public let audioTrackName: String?
}

/// Track reference shared by track lifecycle events.
public struct PlayerTrackEvent: Sendable {
    public let kind: PlayerTrackKind
    public let trackName: String?
    public let epoch: UInt64

    public init(kind: PlayerTrackKind, trackName: String? = nil, epoch: UInt64 = .zero) {
        self.kind = kind
        self.trackName = trackName
        self.epoch = epoch
    }
}

/// Selected-track state after a selection change.
public struct PlayerTrackSelectionEvent: Sendable {
    public let kind: PlayerTrackKind
    public let trackName: String?

    public var isEnabled: Bool {
        trackName != nil
    }
}

/// First accepted/decoded frame for a subscribed track.
public struct PlayerTrackReadyEvent: Sendable {
    public let track: PlayerTrackEvent
    public let sourceTimestampUs: UInt64
    public let targetBuffering: Duration
    public let keyframe: Bool
    public let payloadBytes: UInt64
}

/// First audible or visible playback for a subscribed track.
public struct PlayerTrackPlayingEvent: Sendable {
    public let track: PlayerTrackEvent
    public let sourceTimestampUs: UInt64
    public let targetBuffering: Duration
    public let output: PlayerTrackPlaybackOutput
}

public enum PlayerTrackPlaybackOutput: Sendable {
    case audio(PlayerAudioPlaybackOutput)
    case video(PlayerVideoPlaybackOutput)
}

public struct PlayerAudioPlaybackOutput: Sendable {
    public let timestampUs: UInt64
    public let hostTime: UInt64?
}

public struct PlayerVideoPlaybackOutput: Sendable {
    public let presentationTimeUs: UInt64
    public let clockTimeUs: UInt64
    public let buffer: Duration
}

/// Error associated with a specific track.
public struct PlayerTrackErrorEvent: Sendable {
    public let track: PlayerTrackEvent
    public let message: String
}

/// Playback end context.
public struct PlayerPlaybackEndEvent: Sendable {
    public let reason: String?
}

/// Strongly typed payload for each player event.
public enum PlayerEventType: Sendable {
    case playerInit(PlayerSessionEvent)
    case playerDestroy
    case playbackRequest(PlayerSessionEvent)
    case playbackStart(PlayerTrackPlayingEvent)
    case playbackPause(PlayerSessionEvent)
    case playbackResume(PlayerSessionEvent)
    case playbackEnd(PlayerPlaybackEndEvent)
    case trackSubscribeStart(PlayerTrackEvent)
    case trackReady(PlayerTrackReadyEvent)
    case trackPlaying(PlayerTrackPlayingEvent)
    case trackSubscribeError(PlayerTrackErrorEvent)
    case trackSubscribeEnd(PlayerTrackEvent)
    case trackSelect(PlayerTrackSelectionEvent)
    case trackSwitch(PlayerTrackEvent)
    case trackStallStart(PlayerTrackEvent)
    case trackStallEnd(PlayerTrackEvent)
    case rebufferStart(PlayerTrackEvent)
    case rebufferEnd(PlayerTrackEvent)
    case decodeError(PlayerTrackErrorEvent)

    public var name: PlayerEventName {
        switch self {
        case .playerInit:
            return .playerInit
        case .playerDestroy:
            return .playerDestroy
        case .playbackRequest:
            return .playbackRequest
        case .playbackStart:
            return .playbackStart
        case .playbackPause:
            return .playbackPause
        case .playbackResume:
            return .playbackResume
        case .playbackEnd:
            return .playbackEnd
        case .trackSubscribeStart:
            return .trackSubscribeStart
        case .trackReady:
            return .trackReady
        case .trackPlaying:
            return .trackPlaying
        case .trackSubscribeError:
            return .trackSubscribeError
        case .trackSubscribeEnd:
            return .trackSubscribeEnd
        case .trackSelect:
            return .trackSelect
        case .trackSwitch:
            return .trackSwitch
        case .trackStallStart:
            return .trackStallStart
        case .trackStallEnd:
            return .trackStallEnd
        case .rebufferStart:
            return .rebufferStart
        case .rebufferEnd:
            return .rebufferEnd
        case .decodeError:
            return .decodeError
        }
    }
}

/// A player event envelope.
public struct PlayerEvent: Sendable {
    public let type: PlayerEventType
    public let timestamp: ContinuousClock.Instant
    public let sequence: UInt64

    public var name: PlayerEventName {
        type.name
    }

    init(
        type: PlayerEventType,
        timestamp: ContinuousClock.Instant,
        sequence: UInt64
    ) {
        self.type = type
        self.timestamp = timestamp
        self.sequence = sequence
    }
}

/// A cancellable event or stats subscription.
public final class PlayerEventSubscription: @unchecked Sendable {
    private let lock = NSLock()
    private var onCancel: (() -> Void)?

    init(onCancel: @escaping () -> Void) {
        self.onCancel = onCancel
    }

    deinit {
        cancel()
    }

    public func cancel() {
        let action: (() -> Void)? = lock.withLock {
            let action = onCancel
            onCancel = nil
            return action
        }
        action?()
    }
}

private extension NSLock {
    func withLock<R>(_ body: () -> R) -> R {
        lock()
        defer { unlock() }
        return body()
    }
}

File: android/moqkit/moqkit/src/main/java/com/swmansion/moqkit/subscribe/PlayerEvents.kt
package com.swmansion.moqkit.subscribe

import java.time.Duration
import java.time.Instant

/**
 * Stable event names emitted by [Player.events].
 *
 * Events represent player lifecycle transitions. Periodic quality samples are exposed
 * through [Player.statsUpdates].
 */
enum class PlayerEventName(val value: String) {
    PLAYER_INIT("player.init"),
    PLAYER_DESTROY("player.destroy"),
    PLAYBACK_REQUEST("playback.request"),
    PLAYBACK_START("playback.start"),
    PLAYBACK_PAUSE("playback.pause"),
    PLAYBACK_RESUME("playback.resume"),
    PLAYBACK_END("playback.end"),
    TRACK_SUBSCRIBE_START("track.subscribe.start"),
    TRACK_READY("track.ready"),
    TRACK_PLAYING("track.playing"),
    TRACK_SUBSCRIBE_ERROR("track.subscribe.error"),
    TRACK_SUBSCRIBE_END("track.subscribe.end"),
    TRACK_SELECT("track.select"),
    TRACK_SWITCH("track.switch"),
    TRACK_STALL_START("track.stall.start"),
    TRACK_STALL_END("track.stall.end"),
    REBUFFER_START("rebuffer.start"),
    REBUFFER_END("rebuffer.end"),
    DECODE_ERROR("decode.error"),
}

/** Media kind carried by player events. */
enum class PlayerTrackKind(val value: String) {
    AUDIO("audio"),
    VIDEO("video"),
}

/** Session-level context shared by player lifecycle events. */
data class PlayerSessionEvent(
    val catalogPath: String,
    val targetBuffering: Duration,
    val videoTrackName: String?,
    val audioTrackName: String?,
)

/** Track reference shared by track lifecycle events. */
data class PlayerTrackEvent(
    val kind: PlayerTrackKind,
    val trackName: String? = null,
    val epoch: Long = 0L,
)

/** Selected-track state after a selection change. */
data class PlayerTrackSelectionEvent(
    val kind: PlayerTrackKind,
    val trackName: String?,
) {
    val isEnabled: Boolean
        get() = trackName != null
}

/** First accepted or decoded frame for a subscribed track. */
data class PlayerTrackReadyEvent(
    val track: PlayerTrackEvent,
    val sourceTimestampUs: Long,
    val targetBuffering: Duration,
    val keyframe: Boolean,
    val payloadBytes: Long,
)

/** First audible or visible playback for a subscribed track. */
data class PlayerTrackPlayingEvent(
    val track: PlayerTrackEvent,
    val sourceTimestampUs: Long,
    val targetBuffering: Duration,
    val output: PlayerTrackPlaybackOutput,
)

sealed class PlayerTrackPlaybackOutput {
    data class Audio(val output: PlayerAudioPlaybackOutput) : PlayerTrackPlaybackOutput()
    data class Video(val output: PlayerVideoPlaybackOutput) : PlayerTrackPlaybackOutput()
}

data class PlayerAudioPlaybackOutput(
    val timestampUs: Long,
    val hostTime: Long?,
)

data class PlayerVideoPlaybackOutput(
    val presentationTimeUs: Long,
    val clockTimeUs: Long,
    val buffer: Duration,
)

/** Error associated with a specific track. */
data class PlayerTrackErrorEvent(
    val track: PlayerTrackEvent,
    val message: String,
)

/** Playback end context. */
data class PlayerPlaybackEndEvent(
    val reason: String?,
)

/** Strongly typed payload for each player event. */
sealed class PlayerEventType {
    abstract val name: PlayerEventName

    data class PlayerInit(val session: PlayerSessionEvent) : PlayerEventType() {
        override val name: PlayerEventName = PlayerEventName.PLAYER_INIT
    }

    object PlayerDestroy : PlayerEventType() {
        override val name: PlayerEventName = PlayerEventName.PLAYER_DESTROY
    }

    data class PlaybackRequest(val session: PlayerSessionEvent) : PlayerEventType() {
        override val name: PlayerEventName = PlayerEventName.PLAYBACK_REQUEST
    }

    data class PlaybackStart(val playback: PlayerTrackPlayingEvent) : PlayerEventType() {
        override val name: PlayerEventName = PlayerEventName.PLAYBACK_START
    }

    data class PlaybackPause(val session: PlayerSessionEvent) : PlayerEventType() {
        override val name: PlayerEventName = PlayerEventName.PLAYBACK_PAUSE
    }

    data class PlaybackResume(val session: PlayerSessionEvent) : PlayerEventType() {
        override val name: PlayerEventName = PlayerEventName.PLAYBACK_RESUME
    }

    data class PlaybackEnd(val end: PlayerPlaybackEndEvent) : PlayerEventType() {
        override val name: PlayerEventName = PlayerEventName.PLAYBACK_END
    }

    data class TrackSubscribeStart(val track: PlayerTrackEvent) : PlayerEventType() {
        override val name: PlayerEventName = PlayerEventName.TRACK_SUBSCRIBE_START
    }

    data class TrackReady(val ready: PlayerTrackReadyEvent) : PlayerEventType() {
        override val name: PlayerEventName = PlayerEventName.TRACK_READY
    }

    data class TrackPlaying(val playing: PlayerTrackPlayingEvent) : PlayerEventType() {
        override val name: PlayerEventName = PlayerEventName.TRACK_PLAYING
    }

    data class TrackSubscribeError(val error: PlayerTrackErrorEvent) : PlayerEventType() {
        override val name: PlayerEventName = PlayerEventName.TRACK_SUBSCRIBE_ERROR
    }

    data class TrackSubscribeEnd(val track: PlayerTrackEvent) : PlayerEventType() {
        override val name: PlayerEventName = PlayerEventName.TRACK_SUBSCRIBE_END
    }

    data class TrackSelect(val selection: PlayerTrackSelectionEvent) : PlayerEventType() {
        override val name: PlayerEventName = PlayerEventName.TRACK_SELECT
    }

    data class TrackSwitch(val track: PlayerTrackEvent) : PlayerEventType() {
        override val name: PlayerEventName = PlayerEventName.TRACK_SWITCH
    }

    data class TrackStallStart(val track: PlayerTrackEvent) : PlayerEventType() {
        override val name: PlayerEventName = PlayerEventName.TRACK_STALL_START
    }

    data class TrackStallEnd(val track: PlayerTrackEvent) : PlayerEventType() {
        override val name: PlayerEventName = PlayerEventName.TRACK_STALL_END
    }

    data class RebufferStart(val track: PlayerTrackEvent) : PlayerEventType() {
        override val name: PlayerEventName = PlayerEventName.REBUFFER_START
    }

    data class RebufferEnd(val track: PlayerTrackEvent) : PlayerEventType() {
        override val name: PlayerEventName = PlayerEventName.REBUFFER_END
    }

    data class DecodeError(val error: PlayerTrackErrorEvent) : PlayerEventType() {
        override val name: PlayerEventName = PlayerEventName.DECODE_ERROR
    }
}

/** A player event envelope. */
data class PlayerEvent(
    val type: PlayerEventType,
    val timestamp: Instant,
    val sequence: Long,
) {
    val name: PlayerEventName
        get() = type.name
}
